File: solutions/_402_remove_k_digits/solution.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function (num, k) {
  const stack = [];
  for (const digit of num) {
    while (k > 0 && stack.length > 0 && stack[stack.length - 1] > digit) {
      stack.pop();
      k--;
    }
    if (digit !== "0" || stack.length > 0) {
      stack.push(digit);
    }
  }

  while (k > 0) {
    stack.pop();
    k--;
  }

  return stack.length > 0 ? stack.join("") : "0";
};

module.exports = { removeKdigits };

/*
https://leetcode.com/problems/remove-k-digits/
*/

File: solutions/_402_remove_k_digits/solution.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        for digit in num:
            while stack and stack[-1] > digit and k > 0:
                k -= 1
                stack.pop()
            if stack or digit != "0":
                stack.append(digit)
        if k:
            stack = stack[:-k]
        return "".join(stack) or "0"


# https://leetcode.com/problems/remove-k-digits/
